### ISSUE TITLE
Detect systemd features at runtime, not at build time

### ISSUE THREAD

**Issue by Whissi** [rsyslog/rsyslog#2268]
Currently reviewing why >=systemd-234 is at least highly recommended for >=rsyslog-8.30.

Found commit 83388eabe9a5c5e4b68343c032170fe01363ac45 which looks weird: Why don't you check that at runtime? Users may downgrade systemd without rebuilding systemd (they cannot detect that this is required when linked systemd version is still valid).

**Comment by rgerhards** [comment]
@jvymazal can you please have a look

**Comment by rgerhards** [comment]
@Whissi just a general question: we detect all library versions at compile time. Of course, we could always manually dlload() the libs and check what is available and what not. But does that really make sense? It's a lot of effort and we did not run into issue the last 15 or so years since we do these checks. My understanding also is that autotools were actually designed for this mode of operations. What am I overlooking?

**Comment by jvymazal** [comment]
The "recommended" version of systemd is due to the enabling runtime persistent journal switch (see #1747 ) As Rainer already said I did not want to break compile-time deps checking style present here for years, and the commit you are referring to was added later when we found out that said systemd version is way too new for many rsyslog consumers. It may be taken out after some time as usage of older systemd decreases enough.

**Comment by Whissi** [comment]
Well, in general you are right to use stuff like autotools for deps. But it looks like you are working around a [bug in systemd](https://github.com/systemd/systemd/commit/f934644424daa6c86fd2284fe8f33ea233ece874) (I am not sure if I understood the problem in/with systemd exactly). In this case you shouldn't care. It would be the job of the systemd maintainer in each distribution to backport the fix (you maybe want to ping/convince them to do that ;)). Now you also require a rebuild of rsyslog in case someone backports the fix.

But I am not sure if you can detect that fix at runtime so let's just close the issue. Given that downgrades aren't really supported in most distributions...

**Comment by rgerhards** [comment]
@whissi Let me ask a related question: in 8.32 I was about to require libfastjson 0.99.8 because of the bug fixes it has. In theory, they could be backported but that makes little sense, as it only is a bug fixing release. Nevertheless, would you recommend to leave the requirement at 0.99.7 even though I know this can have very bad effect?

**Comment by Whissi** [comment]
The same bug in libfastjson doesn't affect previous rsyslog version?

**Comment by rgerhards** [comment]
It affects all v8 versions of rsyslog. I hope that distros pick the new
libfastjson version up. But at the same time I want to make sure that the
next rsyslog release does not accept a buggy libfastjson.

**Comment by Whissi** [comment]
So yes, I would recommend to leave the requirement at 0.99.7. Build system should only enforce build requirements. Imagine you would still use json-c: You fixed something in json-c which was merged and the first release containing the fix is json-c-0.13. You would now bump the requirement and require >=json-c-0.13:

This would be a problem for some because json-c-0.13 finally removed [json_object_private.h](https://github.com/json-c/json-c/commit/d447fbca77a8dfca71a849dde2b24411a46ff550) which is breaking a lot of packages.

In the end this will affect adoption rate.

So treat libfastjson like any external library you don't control.

If you contribute a fix use feature branches and squash patches: A single commit fixing one problem makes it easier for package maintainers to cherry-pick.

Back to the systemd workaround: It would be so much easier to backport https://github.com/systemd/systemd/commit/f934644424daa6c86fd2284fe8f33ea233ece874 if this was the problem. And I guess most package maintainers would appreciate such a hint.

**Comment by rgerhards** [comment]
> So treat libfastjson like any external library you don't control.

Sorry for hijacking the thread and being persistent. My intent is not to "be right" but to be sure to really understand best practice here.

> If you contribute a fix use feature branches and squash patches: A single commit fixing one problem makes it easier for package maintainers to cherry-pick.

This is what I do for quite a while. Actually, I usually put the fix in one commit and the new testbench test into the next one, so that it is easier to cherry-pick just the fix.

To libfastjson: have a quick look at this: https://github.com/rsyslog/rsyslog/issues/1839 -- it is a real problem, and I think there is at least one or two other issues with the same root cause. So it is pretty dangerous to use  unpatched libfastjson. In this specific case, libfstjson 0.99.8 is really only a bug-fixing release (from ChangeLog):
```
0.99.8 2017-12-18
- make build under gcc7 with strict settings (warning==error)
- bugfix: constant key names not properly handled
  if fjson_object_object_add_ex() is used with option
  FJSON_OBJECT_KEY_IS_CONSTANT, fjson_object_object_del() will still
  try to delete the key name. Depending on use, this can lead to
  double-free, use-after-free or no problem.
  see also https://github.com/rsyslog/rsyslog/issues/1839
  closes https://github.com/rsyslog/libfastjson/issues/148
- fix potentially invalid return value of fjson_object_iter_begin
  this could lead to callers doing improper opreations and thus
  could lead to a segfault in callers
  detected by Coverity scan, CID 198891
- fix small potential memory leak in json_tokener (unlinkely to occur)
  detected by Coverity Scan, CID 198890
```
I don't see any good reason to backport one specific fix and let the other 3 unfixed. IMHO the right thing to do for a maintainer is just grab the new release and be done. If there were breaking changes, that would be a different beast. But we don't have that (and that's the most frequent case IMHO).

If I keep 0.99.7 valid, I have no way to detect if the critical fix is applied or not (version number check doesn't help). All I could do is query the version number and emit a warning message that libfastjson may miss a patch and rsyslog thus may malfunction. That sounds weird to me. So I would have thought that requiring the fixed version is the way to go (again: I want to understand this, not trying to enforce what I originally intended to do!).

I have a related issue: for kafka components, we wait for a critical bugfix in librdkafka (a memory leak). I would have thought that I increase the version requirement if that version is finally out.

Bottom line: I know some libs to have bad defects for rsyslog (doesn't matter if rsyslog-provided lib or provided by anyone else). So far, I bumbed to version number requirements to the known-good versions of the libs. Should I really leave them at the buggy ones and let user experience hard to track errors if for the old packages no backports happen (and IMHO they usually happen very late, if ever, especially with "enterprise distros").

Again: this is a honest question. I know I do not understand package maintainers policies in full and would like to improve my understanding.

Maybe @mbiebl also would be willing to provide his PoV?

**Comment by mbiebl** [comment]
To understand the libsystemd issue better: Do you actually use a newer symbol from the library which would enforce a stricter runtime dependency. Say libraryA has added a new function methodB in version 2.0. The bumping the version requirement in configure.ac. Debian (and I think Fedora/Redhat) as well, can figure out, that if you use the symbol methodB, the package will get a tighter dependency on libraryA (>= 2.0). 
If the internal behaviour of systemd changed, how do you enforce on binary distros, that a newer systemd version is installed? (runtime != buildtime)

**Comment by rgerhards** [comment]
@jvymazal I would appreciate if you could comment on @mbiebl's question as you probably know exactly why you did it this way - I would need to do a bit of search first.

@mbiebl thx for taking the time to look at this. I would also be especially be interested on your thoughts on libfastjson and librdkafka I mentioned in my last message. This is upcoming to code and I am a bit puzzled on how to do it right. Thx again!

**Comment by jvymazal** [comment]
@mbiebl No, it  was more of a bugfix than new feature so no API change, only internals. 

As for how to solve this on binary distro, configure.ac can help here as well, when you are building what you are going to push to the binary repo, you are building against what is already available in there, so if it is there you can assume users will have it as well. This is as close to "enforcing" as you can get with buildtime and autotools, if you need more strict way you are left with basically using package manager for the distro and specifying the deps and versions there.

**Comment by mbiebl** [comment]
@jvymazal what happens if you builld against  systemd (>= 234) and the user (e.g. by having the systemd package on hold) upgrades rsyslog but has a version << 234 installed?

**Comment by jvymazal** [comment]
That depends whether I used package manger (e.g. dnf in fedora case) to explicitly require versioned dep for rsyslog version in question. If I did and user did not circumvent it (in which case it is his problem and I do not really care) it will force a systemd update with it. If I would be lazy and just hoped that user just updates everything that it is not really defined situation.

**Comment by rgerhards** [comment]
@Whissi @mbiebl sorry for being a pest, but I admit that I am now totally puzzled on what should I do as upstream maintainer to handle this correctly. I still feel bad of knowingly permit building binaries that we know to cause potentially fatal failures...

@davidelang @andreasstieger @janmejay @deoren @flicker581 @portante @taotetek @radu-gheorghe  maybe you would also be willing to dispense some advise?

**Comment by deoren** [comment]
@rgerhards I will defer to the knowledge of others on this one, but will chime in for what it is worth. 

For our systems, we're relying entirely on the packages provided by this team, the upstream providers. In that case I would expect that new releases would have a dependency on (greater than or equals to) the fixed package version to prevent known issues from reoccurring for something you've already fixed.

For downstream maintainers, I am used to seeing them back-port the specific fix to whatever version of the library they're maintaining for their LTS release(s). I believe that this is usually for cases where fixes are introduced upstream alongside behavior/feature changes, but I want to say that I've seen even back-ported bugfixes for bugfix-only upstream releases.

In the case of MySQL packages however, I see Ubuntu just provide the latest upstream version instead of back-porting a rather large pile of fixes from the new release to an older one. I'm not sure if they're just following Debian's lead on that or if it's something specific to their process.

Don't we have a Debian developer/package maintainer who frequents this project? I want to say that I saw some recent ticket activity from them, though I can't recall the ticket at the moment. I think I've also had a Red Hat maintainer chime in and help me before on the mailing list when I've asked questions there. 

I suspect that both would be willing to offer feedback on this.

**Comment by davidelang** [comment]
I think you should depend on the fixed version.

you provide the fixed version of the library so people updating from your repo 
are ok.

any distro sucking in the new version of rsyslog should know that they need to 
pull in the newer version of the library, if it doesn't demand the new version, 
someone may not notice the need.

I don't see a good reason to not depend on the fixed version of a library once 
you make a new release of rsyslog.

**Comment by jvymazal** [comment]
Well, as for fedora, at least in case of rsyslog, we try to stay pretty close to upstream, to avoid dragging already mentioned big pile of patches, as I have said already, if there is some dependency not enforced by auto-tools we force it via packaging system which is the usual in case we need to ensure syncing of bugs/feature concerning more projects/components. 

In case of rsyslog-related stuff we usually group the releases into one update anyway, because there are same maintainers for all of these. So for us, the probably best way is the new/increased deps are clearly said in changelog/release notes/announcement so we know of them and can put them to our spec. 

And for the record I am very glad that there are no runtime deps as that opens whole new landscape of problems which have much more complicated fixes than just ensuring I have correct deps in spec when pushing update.
